extensions/common/iam/identity-trust/identity-trust-service/src/main/java/org/eclipse/edc/iam/identitytrust/verification/JwtPresentationVerifier.java

```diff
@@ -14,18 +14,112 @@
+    private final ObjectMapper objectMapper;
+
+    /**
+     * Verifies the JWT presentation by checking the cryptographic integrity.
+     *
+     * @param jwtVerifier  The JwtVerifier instance used to verify the JWT token.
+     * @param thisAudience The audience to which the token is intended. This must be "this connector's identifier"
+     */
+    JwtPresentationVerifier(JwtVerifier jwtVerifier, String thisAudience, ObjectMapper objectMapper) {
+        this.jwtVerifier = jwtVerifier;
+        this.thisAudience = thisAudience;
+        this.objectMapper = objectMapper;
+    }
+
+
     /**
-     * Computes the cryptographic integrity of a VerifiablePresentation
+     * Verifies the presentation by checking the cryptographic integrity as well as the presence of mandatory claims in the JWT.
      *
-     * @param rawVp The base64-encoded JWT string
+     * @param serializedJwt The serialized JWT presentation to be verified.
+     * @return A Result object representing the verification result, containing specific error messages in case of failure.
      */
-    public Result<Void> verifyPresentation(String rawVp) {
-        throw new UnsupportedOperationException("not yet implemented!");
+    public Result<Void> verifyPresentation(String serializedJwt) {
+
+        // verify the "outer" JWT, i.e. the VP JWT
+        var vpResult = jwtVerifier.verify(serializedJwt, thisAudience);
+        if (vpResult.failed()) {
+            return vpResult;
+        }
+
+        // verify all "inner" VC JWTs
+        try {
+            // obtain the actual VP JSON structure
+            var signedVpJwt = SignedJWT.parse(serializedJwt);
+            var vpClaim = signedVpJwt.getJWTClaimsSet().getClaim(VP_CLAIM);
+            if (vpClaim == null) {
+                return Result.failure("VP-JWT does not contain mandatory claim: " + VP_CLAIM);
+            }
+            var vpJson = vpClaim.toString();
+
+            // obtain the "verifiableCredentials" object inside
+            var map = objectMapper.readValue(vpJson, Map.class);
+            if (!map.containsKey(VERIFIABLE_CREDENTIAL_JSON_KEY)) {
+                return Result.failure("Presentation object did not contain mandatory object: " + VERIFIABLE_CREDENTIAL_JSON_KEY);
+            }
+            var vcTokens = extractCredentials(map.get(VERIFIABLE_CREDENTIAL_JSON_KEY));
+
+            if (vcTokens.isEmpty()) {
+                // todo: this is allowed by the spec, but it is semantic nonsense. Should we return failure or not?
+                return Result.failure("No credential found in presentation.");
+            }
+
+            // every VC is represented as another JWT, so we verify all of them
+            for (String token : vcTokens) {
+                vpResult = vpResult.merge(jwtVerifier.verify(token, signedVpJwt.getJWTClaimsSet().getIssuer()));
+            }
+
+        } catch (ParseException | JsonProcessingException e) {
+            throw new RuntimeException(e);
```

Comment:
Do we really want to throw? Shouldn't we return a failed result instead?

Reply:
I was sort of undecided, but this can only happen, if the key is in an invalid format, or we are not dealing with valid JSON, at which point any further processing would be pointless.